src/openforms/submissions/tasks/emails.py

```diff
@@ -77,14 +63,109 @@ def send_confirmation_email(submission_id: int) -> None:
+                subject=_("Co-sign request for {form_name}").format(
+                    form_name=submission.form.name
+                ),
+                html_body=content,
+                from_email=settings.DEFAULT_FROM_EMAIL,
+                recipient_list=[recipient],
+                text_message=content,
+            )
+        except Exception:
+            logevent.cosigner_email_queuing_failure(submission)
+            raise
+
+        submission.cosign_request_email_sent = True
+        submission.save(update_fields=("cosign_request_email_sent",))
+
+        logevent.cosigner_email_queuing_success(submission)
+
+
+@app.task(
+    base=QueueOnce,
+    ignore_result=True,
+    once={"graceful": True},
+)
+def schedule_emails(submission_id: int) -> None:
+    submission = Submission.objects.get(id=submission_id)
+
+    if (
+        submission.confirmation_email_sent
+        and submission.cosign_request_email_sent
+        and submission.cosign_confirmation_email_sent
+        and submission.payment_complete_confirmation_email_sent
+    ):
+        return
+
+    #
+    # Figure out which email(s) should be sent
+    #
+
+    if submission.waiting_on_cosign and not submission.cosign_request_email_sent:
+        send_email_cosigner.delay(submission_id)
+
+    if not submission.form.send_confirmation_email:
+        logger.debug(
+            "Form %d is configured to not send a confirmation email for submission %d, "
+            "skipping the confirmation e-mail.",
+            submission.form.id,
+            submission.id,
+        )
+        logevent.confirmation_email_skip(submission)
+        return
+
+    execution_options = {}
+    if not submission.confirmation_email_sent:
+        # The confirmation email is sent after the user has submitted a form. It contains the generated PDF report
+        # with information about whether payment/cosign is (still) needed or not. This is scheduled with a 15 countdown
+        # if a payment is needed
+        if submission.payment_required and not submission.payment_user_has_paid:
+            # wait a while and check again
+            execution_options["countdown"] = settings.PAYMENT_CONFIRMATION_EMAIL_TIMEOUT
+
+        send_confirmation_email.apply_async(
+            args=(submission.id,),
+            **execution_options,
+        )
+        return
+
+    # If the confirmation email has already been sent, it may be that the submission was cosigned, the payment was
+    # completed, we are in a registration-retry flow or we are in a payment-status-update-retry flow. The last 2
+    # cases don't need to send a confirmation email.
+    # Edge case: cosign and payment happen at the same time
+    if (
+        not submission.cosign_confirmation_email_sent and submission.cosign_complete
+    ) or (
+        not submission.payment_complete_confirmation_email_sent
+        and submission.payment_user_has_paid
+    ):
+        send_confirmation_email.delay(submission_id)
```

Comment:
does this make sense? You can only end up in this block with the condition `if submission.confirmation_email_sent` (the opposite of L147).

This would lead to the confirmation email being sent more than once, no?

Reply:
My reasoning was that on first completion, the confirmation email is scheduled (with or without countdown).
Then, on payment or on cosign event, a new confirmation email is sent (with or without the cosigner in CC depending on the event).